### ISSUE TITLE
How to model Replay/Drop/Inject?

### ISSUE THREAD

**Issue by mhils** [mitmproxy/mitmproxy#4504]
In #4502 @Prinzhorn brought up that it would be convenient to have a `injected` attribute for WebSocketMessage and TCPMessage. We may get a bit of a convoluted zoo with replay/kill/inject properties, so let's discuss how we want to model this.

## Things we want to model:

 - HTTP Forward request/response ("normal operation")
 - HTTP Client Replay
 - HTTP Server Replay
 - HTTP Request Kill
 - HTTP Response Kill
 - TCP/WS Forward from client to server
 - TCP/WS Forward from server to client
 - TCP/WS Drop from client
 - TCP/WS Drop from server
 - TCP/WS Inject to client
 - TCP/WS Inject to server

High-Level Questions:

 - How can we consistently represent regular forwarding, injection, replay, and dropping across flow types?
 - Do we have any idea what replay could look like for WebSocket flows (and by extension, if replay is per message or per flow?)
 - For TCP and WebSocket flows, do we drop on a message-level (silently discard message), on a flow-level (connection closing), or both? It feels like both could make sense.

## Current Status

```python
Flow.is_replay: Literal["request","response",None]
Flow.error: Error  # a somehwat crude proxy for "killed"
WebSocketMessage.from_client: bool
WebSocketMessage.killed: bool
TCPMessage.from_client: bool
# TCP cannot be killed at the moment apparently ¯\_(ツ)_/¯
```

It's... a bit inconsistent.

# Approach A: Properties

```python
Flow.is_replay: Literal["request","response",None]
Flow.dropped: bool  # or .drop
(TCP|WebSocket)Message.from_client: bool
(TCP|WebSocket)Message.dropped: bool
(TCP|WebSocket)Message.injected: bool
(TCP|WebSocket)Message.is_replay: bool  # ???
```

# Approach B: Properties + Flags

```python
class MessageType(enum.Flag):
    DROPPED = 1
    INJECTED = 2
    REPLAY = 4

Flow.is_replay: Literal["request","response",None]
Flow.dropped: bool
(TCP|WebSocket)Message.from_client: bool
(TCP|WebSocket)Message.type: MessageType
```

(this pretty much is a rehashed version of Approach A with bitflags instead of booleans)

# Approach C: Source/Destination Props

```python
class Source(enum.Enum):
    CLIENT = 1
    SERVER = 2
    MITMPROXY = 3

class Destination(enum.Enum):
    CLIENT = 1
    SERVER = 2
    MITMPROXY = 3
    DROP = 4

Flow.source: Source
Flow.destination: Destination
(TCP|WebSocket)Message.source: Source
(TCP|WebSocket)Message.destination: Destination
```

-------

@mitmproxy/devs, any thoughts on what you prefer? I haven't made my mind up yet, both A and C look good to me in general. Can you spot and problems with either approach?

**Comment by Prinzhorn** [comment]
I'm all-in on C if we can apply it to all use-cases. This also gets rid of the asymmetric `from_client` that always twists my brain. And `Destination = DROP` has some `> /dev/null` vibes to it.

But I would argue that `MITMPROXY` is not specific enough. For example inside a `websocket_message` event this doesn't let me distinguish between a WebSocket message that I myself have just injected via an add-on and one that was replayed.

Could there be one more property that points to the add-on or something? Something that makes `MITMPROXY` more specific without introducing another zoo of enum values. So that I can check if my add-on was the source? This wouldn't work with the simplified add-on syntax (I guess?), but if you need that feature you can easily use a class.
With something like this we could even expose this information in the UI and be like "this add-on injected it" or something like this.

**Comment by mhils** [comment]
> Edit: can `Destination = MITMPROXY` even happen?

Yes, server replay.

> But I would argue that `MITMPROXY` is not specific enough. For example inside a `websocket_message` event this doesn't let me distinguish between a WebSocket message that I myself have just injected via an add-on and one that was replayed.

I think we could distinguish between REPLAY and INJECT. Everything else sounds like overkill, as a last resort you could always resort to the wild west of `flow.metadata`. Is it really important to be more specific here?

**Comment by Prinzhorn** [comment]
> Yes, server replay.

I've never used replay. Isn't server replay `flow.response.Source == MITMPROXY`? Oh I just realized you originally suggested `Flow.source`. Maybe I just found another source of confusion :sweat_smile: . Should the property be on both `request/response`? Say I get a regular request from a browser (`flow.request.Source == CLIENT`). And then I intercept and respond with a `Response.make`. Then this response would implicitly have `flow.response.Source == MITMPROXY`. This might also happen the other way around. E.g. you get a request from the client and you forward it to the server like normal. But inside the `request` event you also "fork"/"clone" the request with a slightly modification (this is great, e.g. for finding IDOR, see https://github.com/nccgroup/AutoRepeater). I don't know if this is something that is currently even possible. Anyway, this fork (basically an ad-hoc replay of a modified version of the same flow) would have `flow.request.Source === MITMPROXY` and `flow.request.Destination == SERVER`.

Am I overthinking again?

> I think we could distinguish between REPLAY and INJECT

Yes that might be enough. The `Response.make` example would be considered INJECT in my eyes.

**Comment by Prinzhorn** [comment]
Ok, maybe we only need `Request.Source` but both `Response.Source` and `Response.Destination`? Because the destination of a request is always the other end?

```
Request.Source == CLIENT # Regular stuff
Request.Source == INJECT # The forking/cloning of a request
Request.Source == REPLAY # Client replay

Response.Source == SERVER # Regular stuff
Response.Source == INJECT # Response.make
Response.Source == REPLAY # Server replay

# Maybe my brain is fried right now, but do we actually need Destination at all?
```

**Comment by Prinzhorn** [comment]
> as a last resort you could always resort to the wild west of `flow.metadata`

This is not possible with the current design for injecting messages. But maybe this brings us on the right track: How about `WebSocketMessage.make` than you can then `flow.websocket.inject`? This might solve #4510 and given the user all the flexibility they want to attach more stuff to their injected message. In addition to `message.Source === INJECT` of course.

**Comment by mhils** [comment]
> Isn't server replay `flow.response.Source == MITMPROXY`?

That's client replay, see
https://docs.mitmproxy.org/stable/overview-features/#client-side-replay and
https://docs.mitmproxy.org/stable/overview-features/#server-side-replay 😉 

Let me expand a bit on Approach C. I don't think we need to move it to Request/Response.

| Flow.source | Flow.dest | Use Case
| --- | --- | ---|
CLIENT | SERVER | Normal forwarding
MITMPROXY | SERVER | HTTP client replay
CLIENT | MITMPROXY | HTTP server replay
CLIENT | DROP | HTTP request kill
DROP | SERVER | HTTP response kill
CLIENT | MITMPROXY | HTTPResponse.make

I do think this is relatively sound. HTTP response killing is the only part that slightly unintuitive. Of course you can split MITMPROXY into REPLAY and INJECT/API, although I don't think it's too useful. A client replay is not much different from an injected flow. In both cases it's a request made up by mitmproxy, usually inspired by a real one.

Let's continue for TCP:

| TCPFlow.source | TCPFlow.dest | Use Case
| --- | --- | --- |
| CLIENT | SERVER | Normal connection.
| CLIENT | DROP | Connection killed before contacting server.
MITMPROXY | SERVER | Yet-to-be-invented TCP client replay
CLIENT | MITMPROXY | Yet-to-be-invented TCP server replay

| TCPMessage.source | TCPMessage.dest | Use Case
| --- | --- | --- |
CLIENT | SERVER | Forward from client to server
SERVER | CLIENT | Forward from server to client
CLIENT | DROP | Drop from client
SERVER | DROP | Drop from server
INJECT | SERVER | Inject to server
INJECT | CLIENT | Inject to client

**Comment by Prinzhorn** [comment]
> > Isn't server replay `flow.response.Source == MITMPROXY`?
> 
> That's client replay, see

Can you please double check what you said, I'm lost :sweat_smile: ? This doesn't make sense in my brain right now. When doing server replay mitmproxy sends stored responses to actual clients. So the `source` of the `response` would be `mitmproxy` right?



> I don't think we need to move it to Request/Response.

Let's see what others comment. To me personally the table you've provided is confusing, I think you have a different mental model than I do. Using `HTTPResponse.make` to create something an then calling this a _destination_ doesn't make sense to me. I think the terms _source_ and _destination_ only make sense on `request` and `response`, so you know where they came from and where they're going. What you have is more like `flow.start` and `flow.end` so you know where/how it started and where/how it ended.

Let's say you replay a request and the response (from an actual server) gets killed. In your table this is:

| Flow.source | Flow.dest | Use Case
| --- | --- | ---|
DROP | SERVER | HTTP response kill

The information where the request was coming from is lost. And like I said above the terminology is not intuitive, how can "drop" be a "source"? What I suggest is:


| request.source | request.destination | response.source | response.destination | Use Case
| ---                             | ---                                    | ---                                | ---                                       | --- 
REPLAY | SERVER | SERVER | DROP | response of a client replay was dropped

`request.source = {client, replay, script}`: where did the request originate (client, replay, script)?
`request.destination = {server, drop}`: was the request forwarded to the server or dropped?
`response.source = {server, replay, script}`: where did the response originate from (server, replay, script)?
`response.destination = {client, drop}`: was the response forwarded to the client or dropped?

I guess with this approach we could also delete the `is_replay` property that says "request" or "response"? Or turn it into a getter that uses `flow.request.source` and `flow.response.source`?

> Of course you can split MITMPROXY into REPLAY and INJECT/API, although I don't think it's too useful. A client replay is not much different from an injected flow. In both cases it's a request made up by mitmproxy, usually inspired by a real one.

I'm not on the fence about having only MITMPROXY (although maybe I prefer SCRIPT?). But I'd prefer having a REPLAY (a core functionality of mitmproxy) separated from SCRIPT (which is more generic, but yes technically REPAY is SCRIPT).

Like I said in the beginning, let's wait for a third party. Maybe @Kriechi has some input?

**Comment by Prinzhorn** [comment]
One more note: in my mental model a single `WebsocketMessage` is on the the same level that either a `HTTPRequest` _or_ `HTTPResponse` is (since it goes in one direction). Which is why I think both need source and destination and not just the parent flow which happens to contain both a request and response.

Ok, just  _one_ more thing: do we also need ERROR? E.g. what if the response wasn't dropped but the underlying connection was killed (e.g. I close my browser mid-response)?

**Comment by mhils** [comment]
> > > >  > Edit: can `Destination = MITMPROXY` even happen?
> > > >  
> > > >  Yes, server replay.
> > > 
> > > Isn't server replay `flow.response.Source == MITMPROXY`?
> > 
> > That's client replay, see
> 
> Can you please double check what you said, I'm lost 😅 ? This doesn't make sense in my brain right now. When doing server replay mitmproxy sends stored responses to actual clients. So the `source` of the `response` would be `mitmproxy` right?

Yes. My original "Yes, server replay." referred to `flow.request.destination == MITMPROXY`, you then baited me with your question into making a wrong statement. 😄 

> Let's say you replay a request and the response (from an actual server) gets killed. In your table this is:

Killing the response from a replayed request doesn't make too much sense, it's never going further than mitmproxy anyways. So you kill at the destination. Unless... you could argue that you kill in `responseheaders` if you don't want to receive the body.

> Ok, just one more thing: do we also need ERROR / ABORT? E.g. what if the response wasn't dropped but the underlying connection was killed (e.g. I close my browser mid-response)?

That is flow.error, I don't think we need to model it here.

-------------

Coming back to the bigger picture, your point about WebSocketMessage being on the same level as HTTPRequest is somewhat convincing. There is a bit of duplication (`request.dest == response.src` should always hold), but that's not too bad. I guess the bigger question is we end up with something that's better than Approach A with that, and I'm almost inclined to say no.
Maybe we just want to keep `Flow.is_replay` and some variant of `Flow.dropped` on the flow level, which is sufficient for HTTP. 

## Approach D

```python
Flow.is_replay: Literal["request","response",None]
Flow.dropped: bool # read-only property for flow.error.msg == KILLED
Flow.drop() # set flow.error = Error(KILLED)
(TCP|WebSocket)Message.source: Source  # see Approach C
(TCP|WebSocket)Message.destination: Destination  # see Approach C
```

**Comment by Prinzhorn** [comment]
> Killing the response from a replayed request doesn't make too much sense, it's never going further than mitmproxy anyways

Sure it doesn't make sense on a higher level. But this assumes the mitmproxy user knows the state of the full system at all times. Example: I replay a batch of flows but a second unrelated add-on makes sure some responses are blocked. The second one doesn't even need to know the source and doesn't care. It just does its job. So if it's technically possible it should also work methinks.

> That is flow.error, I don't think we need to model it here.

You are right, I think I'm mixing two different things here.

> There is a bit of duplication (`request.dest == response.src` should always hold), but that's not too bad

I personally think that's _good_ from a ux standpoint. It's nice to have two ways to access the same thing in different contexts. In the same way that I wouldn't mind `from_client`, `to_client`, `from_server` and `to_server` even thought they're all the same, just because of symmetry and better to read code depending on the context. But it looks like Approach D solves that :+1: 

I think we're just having different points of views. So unless someone else gives some helpful input I'm fine with whatever you think fits best.

**Comment by mhils** [comment]
I think this discussion was a bit wordy but really quite useful. Thanks! 🍰 Looks like we both agree that Approach D solves the problems we had with `(TCP|WebSocket)Message` and at the same time leaves `HTTPFlow` untouched. So we only break APIs where we break them anyways with the next release. 👍 

If someone else has any opinions (@mitmproxy/devs, in particular @Kriechi?) please voice them soon. :)
